op-batcher/batcher/driver.go

```diff
@@ -552,12 +565,56 @@ func (l *BatchSubmitter) cancelBlockingTx(queue *txmgr.Queue[txRef], receiptsCh
 	l.sendTx(txData{}, true, candidate, queue, receiptsCh)
 }
 
+// publishToAltDAAndL1 posts the txdata to the DA Provider and then sends the commitment to L1.
+func (l *BatchSubmitter) publishToAltDAAndL1(txdata txData, queue *txmgr.Queue[txRef], receiptsCh chan txmgr.TxReceipt[txRef], daGroup *errgroup.Group) {
+	// sanity checks
+	if nf := len(txdata.frames); nf != 1 {
+		l.Log.Crit("Unexpected number of frames in calldata tx", "num_frames", nf)
+	}
+	if txdata.asBlob {
+		l.Log.Crit("Unexpected blob txdata with AltDA enabled")
+	}
+
+	// when posting txdata to an external DA Provider, we use a goroutine to avoid blocking the main loop
+	// since it may take a while for the request to return.
+	goroutineSpawned := daGroup.TryGo(func() error {
+		// TODO: probably shouldn't be using the global shutdownCtx here, see https://go.dev/blog/context-and-structs
+		// but sendTransaction receives l.killCtx as an argument, which currently is only canceled after waiting for the main loop
+		// to exit, which would wait on this DA call to finish, which would take a long time.
+		// So we prefer to mimic the behavior of txmgr and cancel all pending DA/txmgr requests when the batcher is stopped.
+		comm, err := l.AltDA.SetInput(l.shutdownCtx, txdata.CallData())
+		if err != nil {
+			l.Log.Error("Failed to post input to Alt DA", "error", err)
+			// requeue frame if we fail to post to the DA Provider so it can be retried
+			// note: this assumes that the da server caches requests, otherwise it might lead to resubmissions of the blobs
+			l.recordFailedDARequest(txdata.ID(), err)
+			return nil
+		}
+		l.Log.Info("Set altda input", "commitment", comm, "tx", txdata.ID())
+		candidate := l.calldataTxCandidate(comm.TxData())
+		l.sendTx(txdata, false, candidate, queue, receiptsCh)
+		return nil
+	})
```

Comment:
I'm trying to think about what might happen if a submission to DA is particularly slow, and the ones behind it are fast.

I'm assuming that all later transactions would be blocked for nonce reasons until the first transaction finishes, and then all the others would unblock, does that sound right? For AltDAs that can handle out of order `SetInput` (like an S3 bucket or something), only the commitment to L1 would be blocked by previous transactions. But some AltDAs would force serialization.

In the cases where AltDA have forced serialization on submission (like a nonce), having one goroutine per transaction doesn't fully parallelize the work, it just moves it out of the main thread, correct?

Reply:
> I'm trying to think about what might happen if a submission to DA is particularly slow, and the ones behind it are fast.

Do you have an example in mind? I can't think of a DA layer where the delta would be meaningful. Unless you're thinking one blob is huge and the other is small for eg?

> I'm assuming that all later transactions would be blocked for nonce reasons until the first transaction finishes, and then all the others would unblock, does that sound right?

Not quite. I added the "open questions" section in the PR description. But the idea right now is that both da errgroup and txmgr errgroup are independent, and hence can have a different goroutine limit. This can be useful for eg if you want to limit eth txs to 5 in parallel (because more than that leads to nonce problems or whatever.. just an eg), whereas you need 100 da goroutines in parallel because they each take 10 mins. This is fine, and the da goroutines will simply block on the txmgr errorgroup to spawn a new txmgr goroutine. DA requests that are faster can also overtake the otherones and end up having their commitment sent to L1 earlier.

> But some AltDAs would force serialization.

Do you have an example in mind? I don't see why this would ever be needed/useful. In my mind, a DA layer should purely be for DA. Ordering should be done at the sequencing layer. Otherwise you run into some weird cases where DA layer could extract MEV, as one example.

> In the cases where AltDA have forced serialization on submission (like a nonce), having one goroutine per transaction doesn't fully parallelize the work, it just moves it out of the main thread, correct?

Still not super sure what you mean by this as I explained above. But let's say the DA layer for some reason requires a nonce to be submitted. You can always use the da server as that serialization coordinator, and have that as a middleware that adds a nonce. So seems like things would be fine even in this model?

Reply:
For examples of a submission which is slow while others are fast: Yeah, data size is a good example, or even just some sort of network issue that disproportionately affects one transaction submission over others.

For an example of a DA with forced serialization: I don't have a concrete example in mind, and I agree that a DA *should* simply serve DA. So maybe there's no issue because I'm imagining a mode of operation that is unrealistic.

I think you're right that things are fine in this model, thanks for thinking about it with me :)